### PR TITLE
NPU disable unused PCIe BAR

### PR DESCRIPTION
**Pull request by msiddare**
NPU device enumeration was failing because of an issue in the cisco
ASIC. This patch will fix address the issue by disabling unused BAR.

Signed-off-by: Madhava Reddy Siddareddygari <msiddare@cisco.com>

### CODE REVIEW

File: patch/cisco-npu-disable-other-bars.patch [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    Original patch message follows:
+
+    For Pacific, only bar 0 is valid. Hence disable all other bars as a work
+    around for spitfire
+
+    CDETS: CSCvi15029
+
+    Signed-off-by: Gaurav Gupta <gauragup@cisco.com>
+    Signed-off-by: Subha Keshavaraj <subkesha@cisco.com>
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:25]: It looks like you just copied the output from `git show`. Please create the patch with `git format-patch …`.

File: patch/cisco-npu-disable-other-bars.patch [hunk 2]
```diff
@@ -0,0 +1,90 @@
+From 2bc0b2e84866a7733fc05a34bd8a93e165db99cc Mon Sep 17 00:00:00 2001
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Tue, 9 Mar 2021 13:54:28 -0800
+Subject: [PATCH] NPU disable other bars
+
+    This is a port of IOS XR patch
+
+        Spitfire-pacific-disable-all-bars-other-than-0.patch
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:8]: Please provide a git commit hash or URL for this patch.

File: patch/cisco-npu-disable-other-bars.patch [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
++#define PCI_DEVICE_ID_LEABA_PALLADIUM  0xa004
++#define PCI_DEVICE_ID_LEABA_ARGON      0xa005
++#define PCI_DEVICE_ID_LEABA_KRYPTON    0xa006
++
++/*
++ * For Pacific A0, only BAR 0 is valid
++ */
++static void silicon_one_fixup(struct pci_dev *dev)
++{
++	int i;
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:51]: This does not follow the Linux coding style. Please check patches with `checkpatch.pl` script in the Linux kernel.

**Review by paulmenzel — COMMENTED** [review]
Thank you for your patches, but reviewing your submission takes too much time because they have so many formal errors. It’d be great, if you got acquainted with the Linux kernel development processes, and followed them.

File: patch/cisco-npu-disable-other-bars.patch (hunk 3 above)
Comment (paulmenzel) [reply patch/cisco-npu-disable-other-bars.patch:38]: Sorry, Firefox display issues in the browser.

File: patch/cisco-npu-disable-other-bars.patch [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Mon, 12 Jul 2021 23:21:08 -0700
+Subject: [PATCH] NPU disable unused PCI BAR
+
+For Cisco ASIC only BAR0 is valid. Not disabling other BAR's
+was resulting in pci_enable_device function failure.
+Further debugging and consultion with Hardware team, issue seems
+to be related to ASIC and workaround suggested is to disable
+unused PCI BAR
+
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:11]: Please also paste the new log messages from one device to the commit message.

File: patch/cisco-npu-disable-other-bars.patch [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
++	int i;
++	struct resource *r;
++
++	for (i = 1; i <= PCI_ROM_RESOURCE; i++) {
++		r = &dev->resource[i];
++		if (!r->start && !r->end && !r->flags)
++			continue;
++
++		dev_info(&dev->dev,
++			"Cisco Silicon One BAR %d %pR fixed up\n", i, r);
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:47]: Maybe:

> Cisco Silicon One BAR %d %pR disabled due to NPU hardware bug

**Comment by msiddare** [comment]
Review comments incorporated and updated series file as per meeting discussion

File: patch/cisco-npu-disable-other-bars.patch [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
++			"NPU hardware bug in P0 Pacific ASIC\n", i, r);
++		r->start = 0;
++		r->end = 0;
++		r->flags = 0;
++	}
++	/*
++	 * Pacific device was misbehaving during rescan not enumerating
++	 * memory for bar. Due to this HW issue, added this workaround
++	 * and verified that during rescan memory gets assigned properly
++	 * to the device. This is only a temporary fix for pacific ASIC
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:57]: Maybe:

> Due to HW bug, Pacific device is misbehaving during rescan by not enumerating memory for that BAR. Work around it and verify, that during rescan memory gets assigned properly to the device.

File: patch/cisco-npu-disable-other-bars.patch [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
++	 * only
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:58]: Is it really temporary?

File: patch/cisco-npu-disable-other-bars.patch [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
++	 */
++	dev->class = PCI_CLASS_MEMORY_OTHER << 8;
++	dev_info(&dev->dev, "Cisco Silicon One class adjusted\n");
++}
++DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_SYNOPSYS, PCI_DEVICE_ID_LEABA_PACIFIC,
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:63]: Why `PCI_VENDOR_ID_SYNOPSYS`?

File: patch/cisco-npu-disable-other-bars.patch [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
++#define PCI_DEVICE_ID_LEABA_GIBRALTAR  0xa001
++#define PCI_DEVICE_ID_LEABA_GRAPHENE   0xa003
++#define PCI_DEVICE_ID_LEABA_PALLADIUM  0xa004
++#define PCI_DEVICE_ID_LEABA_ARGON      0xa005
++#define PCI_DEVICE_ID_LEABA_KRYPTON    0xa006
++
++/*
++ * For Pacific A0, only BAR 0 is valid
++ */
++static void silicon_one_fixup(struct pci_dev *dev)
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:36]: I’d think, the function name needs to include `cisco` somehow.

File: patch/cisco-npu-disable-other-bars.patch [hunk 10]
```diff
@@ -0,0 +1,81 @@
+From 0be339b4c8468138dfe1a30de29511e94b846af6 Mon Sep 17 00:00:00 2001
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Mon, 16 Aug 2021 12:55:37 -0700
+Subject: [PATCH] NPU disable unused PCI BAR
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:4]: Please make it an upstreamable commit message summary ([How to Write a Git Commit Message][1]).

[1]: https://chris.beams.io/posts/git-commit/

File: patch/cisco-npu-disable-other-bars.patch (hunk 4 above)
Comment (paulmenzel) [reply patch/cisco-npu-disable-other-bars.patch:11]: Still open.

File: patch/cisco-npu-disable-other-bars.patch [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+ }
+ DECLARE_PCI_FIXUP_CLASS_HEADER(0x1ac1, 0x089a,
+ 			       PCI_CLASS_NOT_DEFINED, 8, apex_pci_fixup_class);
++
++#define PCI_DEVICE_ID_LEABA_PACIFIC    0xabcd
++#define PCI_DEVICE_ID_LEABA_GIBRALTAR  0xa001
++#define PCI_DEVICE_ID_LEABA_GRAPHENE   0xa003
++#define PCI_DEVICE_ID_LEABA_PALLADIUM  0xa004
++#define PCI_DEVICE_ID_LEABA_ARGON      0xa005
++#define PCI_DEVICE_ID_LEABA_KRYPTON    0xa006
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:31]: Should the macros go in some header file?

File: patch/cisco-npu-disable-other-bars.patch [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
++
++/*
++ * For Pacific A0, only BAR 0 is valid
```
Comment (paulmenzel) [inline patch/cisco-npu-disable-other-bars.patch:34]: Please add an errata document name and number for reference.

**Review by paulmenzel — CHANGES_REQUESTED** [review]
Please get it reviewed by the Linux subsystem maintainers.

**Comment by msiddare** [comment]
Closing this PR as issue is observed only on early P0 proto boards.

**Pull request closed without merging**